File: 021. Palindrome Check/PalindromeCheckOnOnStack.java
import java.util.*;

class PalindromeCheckOnOnStack {
    public static boolean isPalindrome(String str) {
        int length = str.length();
        if (length == 0) {
            return false;
        } else if (length == 1) {
            return true;
        }

        char currentChar;
        boolean isOdd = (length % 2 == 0) ? false : true;
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < length; i++) {
            if (isOdd) {
                if (i == length / 2) {
                    // when current index is the centre pivot
                    continue;
                } else if (i < length / 2) {
                    stack.push(str.charAt(i));
                } else {
                    currentChar = stack.pop();
                    if (currentChar != str.charAt(i)) {
                        return false;
                    }
                }
            } else if (!isOdd) {
                if (i < length / 2) {
                    stack.push(str.charAt(i));
                } else {
                    currentChar = stack.pop();
                    if (currentChar != str.charAt(i)) {
                        return false;
                    }
                }
            }
        }
        return (stack.size() == 0) ? true : false;
    }
}

File: 021. Palindrome Check/PalindromeCheckOnOnStack.kt
package com.algoexpert.program

import java.util.ArrayDeque

fun PalindromeCheckOnOnStack(string: String): Boolean {
    val length = string.length
    if (length == 0) {
        return false
    } else if (length == 1) {
        return true
    }

    var currentChar: Char
    val isOdd: Boolean = if (length % 2 == 0) false else true
    var stack = ArrayDeque<Char>()
    for (i in 0 until length) {
        if (isOdd) {
            if (i == length / 2) {
                // when current index is the centre pivot
                continue
            } else if (i < length / 2) {
                stack.push(string.get(i))
            } else {
                currentChar = stack.pop()
                if (currentChar != string.get(i)) {
                    return false
                }
            }
        } else {
            if (i < length / 2) {
                stack.push(string.get(i))
            } else {
                currentChar = stack.pop()
                if (currentChar != string.get(i)) {
                    return false
                }
            }
        }
    }
    return if (stack.size == 0) true else false
}
